python/dtype.h

```diff
@@ -13,8 +13,39 @@ namespace pybind11 {
 class dtype;
 }
 
+scipp::core::DType dtype_of(const pybind11::object &x);
+
 scipp::core::DType scipp_dtype(const pybind11::object &type);
 
+void ensure_conversion_possible(scipp::core::DType from, scipp::core::DType to,
+                                const std::string &data_name);
+
+template <class T, class = void> struct converting_cast {
+  static decltype(auto) cast(const pybind11::object &obj) {
+    return obj.cast<T>();
+  }
+};
+
+template <class T>
+struct converting_cast<T, std::enable_if_t<std::is_integral_v<T>>> {
+  static decltype(auto) cast(const pybind11::object &obj) {
+    if (dtype_of(obj) == scipp::dtype<double>) {
+      // This conversion is not implemented in pybind11 v2.6.2
+      return obj.cast<pybind11::int_>().cast<T>();
```

Comment:
Yes. This was already possible before because pybind11 handled it during overload resolution. I really don't know why this is the only case where `cast` does not work without the intermediate step.